### PR TITLE
Experimental object creation for Databricks statements

### PR DESCRIPTION
**Pull request by Sondergaard**
## Description

First iteration of object creation for Databricks statements.

~~This is blocked by PR #272~~

## Quality

- [x] Documentation is updated
- [x] Release notes are updated
- [x] Package version is updated
- [x] Public types and methods are documented
- [x] Tests are implemented and executed locally

### CODE REVIEW

File: source/Databricks/source/SqlStatementExecution/DatabricksSqlWarehouseQueryExecutor.cs [hunk 1]
```diff
@@ -91,6 +91,46 @@ await foreach (var record in DoExecuteStatementAsync(statement, format))
...
+
+            if (chunkResponse?.external_links == null) continue;
+
+            await using var stream = await _externalHttpClient.GetStreamAsync(chunkResponse.external_links[0].external_link);
+
+            await foreach (var row in strategy.ExecuteAsync<T>(stream))
+            {
+                yield return row;
+            }
+        }
```
Comment (HenrikSommer) [inline source/Databricks/source/SqlStatementExecution/DatabricksSqlWarehouseQueryExecutor.cs:131]: L#107 - L#131 is duplicated code, maybe we should put that in a private method?

File: source/Databricks/source/SqlStatementExecution/Formats/Reflections.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+internal static class Reflections
+{
+    public static string[] GetArrowFieldNames<T>()
+        => Column<T>._names;
+
+    public static T CreateInstance<T>(params object?[] values)
+        => Creator<T>._create(values);
+
+    private static class Column<T>
```
Comment (dstenroejl) [inline source/Databricks/source/SqlStatementExecution/Formats/Reflections.cs:30]: I don't get the meaning of this class name. Would it be possible to find a more meaningful name?

File: source/Databricks/source/SqlStatementExecution/Formats/Reflections.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        public static readonly string[] _names = GetNames(typeof(T));
+
+        private static string[] GetNames(Type type)
```
Comment (dstenroejl) [inline source/Databricks/source/SqlStatementExecution/Formats/Reflections.cs:34]: Suggestion: Use a descriptive name like `GetNamesByConstructorOrder` or just `GetNamesByOrder`

File: source/Databricks/source/SqlStatementExecution.IntegrationTests/Client/ObjectCreationTests.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Energinet.DataHub.Core.Databricks.SqlStatementExecution.Formats;
+using Energinet.DataHub.Core.Databricks.SqlStatementExecution.IntegrationTests.Fixtures;
+using FluentAssertions;
+
+namespace Energinet.DataHub.Core.Databricks.SqlStatementExecution.IntegrationTests.Client;
+
+public class ObjectCreationTests : IClassFixture<DatabricksSqlWarehouseFixture>
```
Comment (dstenroejl) [inline source/Databricks/source/SqlStatementExecution.IntegrationTests/Client/ObjectCreationTests.cs:21]: Suggestion: Add a test showing a non-sunshine scenario to document expected behavior to developers. E.g. incorrect parameter order or type in constructor.

File: source/Databricks/source/SqlStatementExecution/DatabricksSqlWarehouseQueryExecutor.cs [hunk 5]
```diff
@@ -91,6 +91,46 @@ await foreach (var record in DoExecuteStatementAsync(statement, format))
         }
     }
 
+    /// <summary>
+    /// Asynchronously executes a parameterized SQL query on Databricks and streams the result back as a collection of strongly typed objects.
+    /// </summary>
+    /// <param name="statement">The SQL query to be executed, with collection of <see cref="QueryParameter"/> parameters.</param>
+    /// <typeparam name="T">Target type</typeparam>
```
Comment (dstenroejl) [inline source/Databricks/source/SqlStatementExecution/DatabricksSqlWarehouseQueryExecutor.cs:98]: I believe there are some constraints to the type that we could perhaps document here. E.g. the type must use the attribute, and it must be a record or at least have a constructor with parameters matching the fields.

**Review by dstenroejl — COMMENTED** [review]
Added a few suggestions

File: source/Databricks/source/SqlStatementExecution/DatabricksSqlWarehouseQueryExecutor.cs (hunk 1 above)
Comment (Sondergaard) [reply source/Databricks/source/SqlStatementExecution/DatabricksSqlWarehouseQueryExecutor.cs:131]: The loop-part in line 127 > 130 is different. The row variable is of type `T` where as in the other method the variable is of type `dynamic`.

I have tried to move it into a separate function but it didn't work - I think the code might need to be restructured to reduce duplication.

File: source/Databricks/documents/release-notes/release-notes.md [hunk 6]
```diff
@@ -1,5 +1,13 @@
 # Databricks Release Notes
 
+## Version 8.1.0
+
+- Create objects from statements as an alternative to dynamic
+
+## Version 8.02
```
[the remark this answers is not in the document]
Comment (Sondergaard) [reply source/Databricks/documents/release-notes/release-notes.md]: Updated

File: source/Databricks/source/SqlStatementExecution/Formats/Reflections.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+            return fields.OrderBy(f => f.ConstructorOrder).Select(f => f.Name).ToArray();
+        }
+    }
+
+    private static class Creator<T>
+    {
+        public static readonly Func<object?[], T> _create = CreateCreator();
+
+        private static Func<object?[], T> CreateCreator()
```
[the remark this answers is not in the document]
Comment (Sondergaard) [reply source/Databricks/source/SqlStatementExecution/Formats/Reflections.cs]: Method has been renamed

File: source/Databricks/source/SqlStatementExecution/Formats/Reflections.cs (hunk 2 above)
Comment (Sondergaard) [reply source/Databricks/source/SqlStatementExecution/Formats/Reflections.cs]: Class renamed

File: source/Databricks/source/SqlStatementExecution/Formats/Reflections.cs (hunk 3 above)
Comment (Sondergaard) [reply source/Databricks/source/SqlStatementExecution/Formats/Reflections.cs]: Method renamed to a more descriptive name

File: source/Databricks/source/SqlStatementExecution.IntegrationTests/Client/ObjectCreationTests.cs (hunk 4 above)
Comment (Sondergaard) [reply source/Databricks/source/SqlStatementExecution.IntegrationTests/Client/ObjectCreationTests.cs:21]: Clouds are darkish grey, the wind starts to pick up followed by rain - this mood is now reflected in tests

File: source/Databricks/source/SqlStatementExecution/DatabricksSqlWarehouseQueryExecutor.cs (hunk 5 above)
Comment (Sondergaard) [reply source/Databricks/source/SqlStatementExecution/DatabricksSqlWarehouseQueryExecutor.cs:98]: Documentation updated

File: source/Databricks/source/SqlStatementExecution.IntegrationTests/Client/ObjectCreationTests.cs (hunk 4 above)
Comment (dstenroejl) [reply source/Databricks/source/SqlStatementExecution.IntegrationTests/Client/ObjectCreationTests.cs:21]: Love the class names ;)

File: source/Databricks/source/SqlStatementExecution.IntegrationTests/Client/ObjectCreationTests.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        // Act
+        var result = client.ExecuteStatementAsync<BadPerson>(PersonsStatement);
+        Func<Task> act = async () => await result.ToListAsync();
+
+        // Assert
+        await act.Should().ThrowAsync<InvalidOperationException>();
+    }
+
+    [Fact]
+    public async Task GivenAClassWithToParameters_WhenOnlyOneIsMapped_ThenExceptionIsThrown()
```
Comment (dstenroejl) [inline source/Databricks/source/SqlStatementExecution.IntegrationTests/Client/ObjectCreationTests.cs:68]:
```suggestion
    public async Task GivenAClassWithTwoParameters_WhenOnlyOneIsMapped_ThenExceptionIsThrown()
```

File: source/Databricks/source/SqlStatementExecution/DatabricksSqlWarehouseQueryExecutor.cs (hunk 5 above)
Comment (dstenroejl) [reply source/Databricks/source/SqlStatementExecution/DatabricksSqlWarehouseQueryExecutor.cs:98]: Nice

**Review by dstenroejl — APPROVED** [review]
LGTM

**Review by HenrikSommer — APPROVED** [review]
I have committed Dan's test name suggestion and approved.

**Pull request merged**